### PR TITLE
Fixed #35074 -- Fixed adding/removing indexes when spatial_index is changed on MySQL, PostgreSQL, and Oracle.

### PR DESCRIPTION
**Pull request by felixxm**
ticket-35074

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
I'd keep commits separately.

File: django/contrib/gis/db/backends/postgis/schema.py [hunk 1]
```diff
@@ -82,3 +60,31 @@ def _alter_column_type_sql(
...
+            fields=fields,
+            name=name,
+            using=" USING %s" % self.geom_index_type,
+            opclasses=opclasses,
+            expressions=expressions,
+        )
+
+    def _delete_spatial_index_sql(self, model, field):
+        index_name = self._create_spatial_index_name(model, field)
+        return self._delete_index_sql(model, index_name)
```
Comment (sarahboyce) [inline django/contrib/gis/db/backends/postgis/schema.py:124]: I think this is unused at this point (i.e by the second commit) but gets used in the next/third commit: c4dfc842435ed2eee47d6f6d5ed3f1f6b3bb3771
How do you feel if I move this `_delete_spatial_index_sql` into the third commit? 

It's less symmetrical than being added here, but I'm hoping to merge in the first two commits and this makes it feel slightly more coupled to the third commit

Comment (felixxm) [reply django/contrib/gis/db/backends/postgis/schema.py:124]: Both works for me :+1:

File: django/contrib/gis/db/backends/postgis/schema.py [hunk 2]
```diff
@@ -58,6 +59,39 @@ def _alter_column_type_sql(
...
+        old_field_spatial_index = (
+            isinstance(old_field, GeometryField) and old_field.spatial_index
+        )
+        new_field_spatial_index = (
+            isinstance(new_field, GeometryField) and new_field.spatial_index
+        )
+        if not old_field_spatial_index and new_field_spatial_index:
+            self.execute(self._create_spatial_index_sql(model, new_field))
+        elif old_field_spatial_index and not new_field_spatial_index:
+            self.execute(self._delete_spatial_index_sql(model, old_field))
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/contrib/gis/db/backends/postgis/schema.py:93]: I added a test for altering field with a spatial index (without changing it).

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/contrib/gis/db/backends/mysql/schema.py [hunk 3]
```diff
@@ -58,6 +58,43 @@ def remove_field(self, model, field):
...
+        old_field_spatial_index = (
+            isinstance(old_field, GeometryField)
+            and old_field.spatial_index
+            and not old_field.null
+        )
+        new_field_spatial_index = (
+            isinstance(new_field, GeometryField)
+            and new_field.spatial_index
+            and not new_field.null
+        )
```
Comment (sarahboyce) [inline django/contrib/gis/db/backends/mysql/schema.py:92]: I'm curious why this one is different to the others, when I remove the `null` checks I didn't get failures locally. Did I miss something?
```suggestion
        old_field_spatial_index = (
            isinstance(old_field, GeometryField) and old_field.spatial_index
        )
        new_field_spatial_index = (
            isinstance(new_field, GeometryField) and new_field.spatial_index
        )
```

File: django/contrib/gis/db/backends/postgis/schema.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarahboyce) [reply django/contrib/gis/db/backends/postgis/schema.py:93]: This looks great thank you

**Review by sarahboyce — COMMENTED** [review]
Thank you @felixxm :star:
I have one last question and then I think we're good to go :+1:

File: django/contrib/gis/db/backends/mysql/schema.py (hunk 3 above)
Comment (felixxm) [reply django/contrib/gis/db/backends/mysql/schema.py:92]: > Did I miss something?

Yes, MySQL doesn't support spatial indexes on `NULL` columns (20c2d625d3d5062e43918d1d7b6f623202491dd4). I can add MySQL-specific test for altering nullable field, if you think it's needed.

Comment (sarahboyce) [reply django/contrib/gis/db/backends/mysql/schema.py:92]: Ah thank you
> I can add MySQL-specific test for altering nullable field, if you think it's needed.

If it's simple and you don't mind, then yes please, but don't worry if it's complex or awkward :+1: 
Apparently I even looked at that commit but have the memory of a goldfish.

Comment (felixxm) [reply django/contrib/gis/db/backends/mysql/schema.py:92]: np, I will do this later today. I don't have my "Django" laptop with me :smile:

Comment (felixxm) [reply django/contrib/gis/db/backends/mysql/schema.py:92]: Added, sorry for being so slow :sweat_smile:

Comment (sarahboyce) [reply django/contrib/gis/db/backends/mysql/schema.py:92]: Brilliant thank you

**Review by sarahboyce — APPROVED** [review]
Thank you this looks great

**Comment by sarahboyce** [comment]
buildbot, test on oracle.

**Pull request merged**